spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveIcebergProcedures.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.CatalystTypeConverters
+import org.apache.spark.sql.catalyst.expressions.{Expression, Literal}
+import org.apache.spark.sql.catalyst.plans.logical.{Call, CallArgument, LogicalPlan}
+import org.apache.spark.sql.catalyst.procedures.{IcebergProcedureRegistry, OptionalParameter, Parameter, Procedure}
+import org.apache.spark.sql.catalyst.rules.Rule
+import scala.collection.mutable
+
+// TODO: case sensitivity?
+object ResolveIcebergProcedures extends Rule[LogicalPlan] {
```

Comment:
Is it possible to split the extension into features we'd like to ideally have upstream and parts for Iceberg? It would be nice if this were generic so that we could call any stored procedure from a catalog and Iceberg's catalog injects the stored procedures.

Is that possible without a ton of work? We might have to keep the interfaces in the spark3 module.

Reply:
I guess we can try to think how it can look like in Spark. For example, we may have something like this:

```
public interface ProcedureCatalog extends CatalogPlugin {
  Procedure loadProcedure(Identifier ident);
}

public interface Procedure {
  String name();
  Parameter[] parameters();
  MethodHandle methodHandle();
}
```

We will have to keep these in `org.apache.spark.sql.connector.catalog` in Spark3 and implement in Iceberg catalogs.

Then we can have generic rules for procedure resolution in spark3-extensions.

@rdblue, is this what you mean?

Reply:
That would make sense to me, That way there would be a completely separable "Procedure" capability for catalog

Reply:
Yeah, something like that would make sense. I'd like to make sure we can move this upstream as easily as possible, so making it generic like you suggest would be great.

Reply:
I've updated to make it generic. It is still in a rough state but is reviewable.